lib/teiserver/account/libs/user_totp_lib.ex

```diff
@@ -0,0 +1,224 @@
+  @allowed_invalid_attempts 5
+
+  @spec get_user_totp(integer) :: {:inactive, nil} | {:active, TOTP.t()}
+  defp get_user_totp(user_id) do
+    case Repo.get_by(TOTP, user_id: user_id) do
+      nil ->
+        {:inactive, nil}
+
+      totp ->
+        {:active, totp}
+    end
+  end
+
+  @spec get_user_totp_status(integer) :: :active | :inactive
+  def get_user_totp_status(user_id) do
+    {status, _totp} = get_user_totp(user_id)
+    status
+  end
+
+  @spec get_account_locked(integer) :: boolean
+  def get_account_locked(user_id) do
+    {_status, totp} = get_user_totp(user_id)
+
+    cond do
+      is_nil(totp) ->
+        false
+
+      totp.wrong_otp < @allowed_invalid_attempts ->
+        false
+
+      totp.wrong_otp >= @allowed_invalid_attempts ->
+        true
+    end
+  end
+
+  @spec get_or_generate_secret(integer) :: {:new | :existing, binary()}
+  def get_or_generate_secret(user_id) do
+    case get_user_totp(user_id) do
+      {:inactive, nil} ->
+        {:new, NimbleTOTP.secret()}
+
+      {:active, totp} ->
+        {:existing, totp.secret}
+    end
+  end
+
+  @spec get_user_secret(integer) :: binary | nil
+  def get_user_secret(user_id) do
+    case get_user_totp(user_id) do
+      {:inactive, nil} ->
+        :inactive
+
+      {:active, totp} ->
+        totp.secret
+    end
+  end
+
+  @spec get_last_used_otp(integer) :: :inactive | DateTime.t()
+  def get_last_used_otp(user_id) do
+    case get_user_totp(user_id) do
+      {:inactive, nil} ->
+        :inactive
+
+      {:active, totp} ->
+        totp.last_used
+    end
+  end
+
+  @spec set_totp(nil, map()) ::
+          {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()}
+  defp set_totp(nil, _attrs) do
+    changeset = TOTP.changeset(%TOTP{}, %{})
+    {:error, %{changeset | errors: [user: {"user must be persisted", []}]}}
+  end
+
+  @spec set_totp(TOTP.t(), map()) ::
+          {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()}
+  defp set_totp(%TOTP{} = totp, attrs) do
+    totp
+    |> TOTP.changeset(attrs)
+    |> Repo.update()
+  end
+
+  @spec set_totp(integer, map()) ::
+          {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()}
+  defp set_totp(user_id, attrs) do
+    case get_user_totp(user_id) do
+      {:inactive, nil} ->
+        TOTP.changeset(%TOTP{}, attrs)
+        |> Repo.insert()
+
+      {:active, totp} ->
+        set_totp(totp, attrs)
+    end
+  end
+
+  @spec set_secret(integer, binary) :: {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()}
+  def set_secret(user_id, secret) do
+    set_totp(user_id, %{user_id: user_id, secret: secret})
+  end
+
+  @spec set_last_used(User.t(), integer) ::
+          {:ok, TOTP.t()} | {:error, Ecto.Changeset.t()} | {:inactive, User.t()}
+  def set_last_used(%User{id: user_id} = user, last_used) do
+    if get_user_totp_status(user_id) == :active do
+      {:ok, utc_datetime} = DateTime.from_unix(last_used, :second)
+      set_totp(user_id, %{user_id: user_id, last_used: utc_datetime})
+    else
+      {:inactive, user}
```

Comment:
Why is this returning the user? Seems to be used in tests only, should just return `:inactive` here and the new `last_used` can be grabbed in tests if needed. This function could then accept just `user_id`.